src/nvme/mi.c

```diff
@@ -812,6 +865,38 @@ int nvme_mi_admin_identify_partial(nvme_mi_ctrl_t ctrl,
 	return 0;
 }
 
+int nvme_mi_service_primitive_control(nvme_mi_ep_t ep, __u8 opcode, __u16 *result_cpsr)
+{
+	struct nvme_mi_service_resp_hdr resp_hdr;
+	struct nvme_mi_service_req_hdr req_hdr;
+	struct nvme_mi_resp resp;
+	struct nvme_mi_req req;
+	int rc = 0;
+	__le16 cpsr = 0xffff;
+
+	nvme_mi_service_init_req(&req, &req_hdr, opcode);
+
+	nvme_mi_calc_req_mic(&req);
+
+	nvme_mi_service_init_resp(&resp, &resp_hdr);
+
+	rc = nvme_mi_submit(ep, &req, &resp);
+	if (rc)
+		return rc;
+
+	rc = nvme_mi_service_parse_status(&resp, &cpsr);
+	if (rc)
+		return rc;
+
+	if (result_cpsr)
+		*result_cpsr = le16_to_cpu(cpsr);
```

Comment:
You're double-handling the cpsr field here. Probably best to perform the endian conversion in _parse_status, and set the cspr value even on non-zero status.

Reply:
okay, done